[Codegen][GPU] Support all kinds in combiningKindToAllReduce()

andfau-amd opened the pull request:
Changes this function so it supports all members of the CombiningKind enum; the min/max discrepancies that originally motivated their exclusion no longer exist. Since it covers all enum members, it is changed to no longer return an optional. Note that this changes the generated MLIR (gpu.subgroup_reduce is generated), but that should ultimately lower to the same pattern of shuffles as before.

Function is made `static` as it is not used anywhere else and has no prototype in a header.

kuhar commented on compiler/src/iree/compiler/Codegen/Utils/GPUUtils.h:
```diff
@@ -121,6 +122,11 @@ void propagateSharedMemoryCopy(mlir::FunctionOpInterface funcOp);
+/// Convert reduction operation from the Vector dialect enum to the GPU enum.
+/// The reverse conversion is mlir::gpu::convertReductionKind.
+mlir::gpu::AllReduceOperation
+combiningKindToAllReduce(mlir::vector::CombiningKind kind);
```
Should we keep it a static function for now since there are no other uses outside of `emitGPUGroupReduction`?

andfau-amd replied:
I could make the header change in a subsequent commit where I start using it, then? That's fine with me.

kuhar replied:
Ah, then to me it's fine to keep as-is.

kuhar reviewed (approved):
LGTM

Hardcode84 commented on compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp:
```diff
@@ -522,28 +522,31 @@ Value getCombiningIdentityValue(Location loc, OpBuilder &builder,
-    break;
-  }
-  return std::nullopt;
+  switch (unsigned(kind)) {
```
Why do we need unsigned cast?

andfau-amd replied:
I think this was originally because I was getting some weird compiler warning, but removing the casts doesn't stop it building, so I'll do that.

andfau-amd committed:
[Codegen][GPU] Support all kinds in combiningKindToAllReduce()

Changes this function so it supports all members of the CombiningKind
enum; the min/max discrepancies that originally motivated their
exclusion no longer exist. Since it covers all enum members, it is
changed to no longer return an optional. Note that this changes the
generated MLIR (gpu.subgroup_reduce is generated), but that should
ultimately lower to the same pattern of shuffles as before.

Function is made `static` as it is not used anywhere else and has no
prototype in a header.

Signed-off-by: Andrea Faulds <andrea.faulds@amd.com>
